### ISSUE TITLE
Replace RATT with RocketRML

### ISSUE THREAD

**Issue by pietervaneverdingen** [pldn/LDWizard#54]
### User story

- [Who] As maintainer of LD Wizard,
- [What] I want to use [RocketRML](https://github.com/semantifyit/RocketRML) for the data transformation step,
- [Why] So that I can get rid of the currently used 'ratt' dependency.

### Background

- LD Wizard is an application that works in the browser; without a server component.
- LD Wizard currently uses a dependency called 'ratt' to transform the data.
- Moving forward, LD Wizard wants to use an RML native transformation tool.
- RocketRML is (the only?) RML transformation tool that works in the web browser.

### Acceptance criteria

- The 'ratt' dependency is removed from the codebase.
- The transformation works in the same way, but now using RocketRML.
- Verification that the transformations still works with two or three example CSV files.

---

A [bounty](https://www.pldn.nl/wiki/Bounties_List) has been placed on this issue by [PLDN](https://www.pldn.nl) for €1000
 
[Click here to learn more](https://www.pldn.nl/wiki/Bounty_Procedures) if you're interested in claiming this bounty by resolving this issue.

**Comment by vemonet** [comment]
Hi everyone!

I replaced RATT by RocketRML in the LDWizard, it is implemented in this branch of my fork: https://github.com/vemonet/LDWizard/tree/add-rocketrml

* I removed the `@triply/ratt` dependency from the `package.json` and all code related to RATT (more than 3.000 lines deleted! 🎉). The code to perform the transformation with RocketRML is much more succinct than RATT, which will be easier to maintain in the future.
* The transformation works exactly as before, as far as I tried, IRI column refinements are working. I just need to test custom column refinements, but we first need to fix an issue to do so: https://github.com/pldn/LDWizard/issues/80
* The RATT script download has been removed. It was requested in https://github.com/pldn/LDWizard/issues/58 I also removed the support for CSVw/COW and updated the LDWizard text accordingly. 
  * But I think it was a nice feature to share the description in another open machine-readable format that is more geared around describing the CSV than converting it to RDF! (even if we don't make sure it can be used by CoW for transformation)

* I needed to make some fix to the RocketRML library to make it work, mainly to:

  *  support CSVw objects as `rml:source` (cf. https://github.com/semantifyit/RocketRML/issues/40)
  * comment a problematic optional import (cf. https://github.com/semantifyit/RocketRML/issues/39). 

  Hence the RocketRML package is currently imported from a fork branch: https://github.com/vemonet/RocketRML/tree/no-cpp-xml-parser

You can test the LDWizard with RocketRML by cloning my fork, going to the `add-rocketrml` branch, installing, and starting the LDWizard in development:

```bash
git clone https://github.com/vemonet/LDWizard
cd LDWizard
git checkout add-rocketrml
yarn
yarn dev
```

Let me know if you have any comment or see any improvements to do!

Here is what still need to be done from my point of view:

* Test with custom column refinements, once they have been fixed
  * https://github.com/pldn/LDWizard/issues/80
* Ideally we should fix the issues with RocketRML to be able to use the version published to npm instead of my fork 
  * https://github.com/semantifyit/RocketRML/issues/40 
  * https://github.com/semantifyit/RocketRML/issues/39
* Ideally we should also fix the issue with deploying the LDWizard from custom config, so we can try it the same way users will use it 
  * https://github.com/pldn/LDWizard/issues/79
* There are no "official" CSV files to use to check if the LDWizard works as expected. That would be nice to have a few examples CSV provided in the repository, with different LDWizard config to test various features of the LDWizard (I say it's nice to have, but it is an euphemism, if you want the LDWizard to be maintainable you absolutely need testing). Ideally, in the future, those tests could be automatized (at least just to check if the LDWizard test configs are loading without error). 
  * The lack of tests has already been mentioned here: https://github.com/pldn/LDWizard/issues/38

**Comment by vemonet** [comment]
Update: the error with column refinements has been fixed so I could try RocketRML with custom refinements and it works fine

So all is good!

**Comment by pmaria** [comment]
With regards to the CSVW sources. In an earlier meeting we had already discussed that this is currently not supported by RocketRML.

For now we can make due with a plain CSV source. The specific delimiter can be provided through configuration.
Of course, if support is added in RocketRML we can convert tot that.

**Comment by pmaria** [comment]
@vemonet does my comment make sense? Do we need to discuss?

**Comment by vemonet** [comment]
@pmaria I think, so that means we need to change the RML mappings produced by the LDWizard to not use CSVw? (to use just a string for `rml:source` instead of a `csvw:Table` object)

Otherwise I already made the change to add support for CSVw to RocketRML, mentionned in https://github.com/semantifyit/RocketRML/issues/40, so if RocketRML is fine to integrate them it would be quick, and we would not need to change how the source are defined

**Comment by pmaria** [comment]
@vemonet 

> @pmaria I think, so that means we need to change the RML mappings produced by the LDWizard to not use CSVw? (to use just a string for `rml:source` instead of a `csvw:Table` object)

Indeed.
 
> Otherwise I already made the change to add support for CSVw to RocketRML, mentionned in [semantifyit/RocketRML#40](https://github.com/semantifyit/RocketRML/issues/40), so if RocketRML is fine to integrate them it would be quick, and we would not need to change how the source are defined in the LDWizard

Right, I can imagine however, that to fully support CSVW (for example also supporting `csvw:Dialect`, wherein the main selling point of CSVW lies IMHO), more work will be necessary. I think we shouldn't wait for this.
Additionally, the KG-construct working group is also working on new specifications which might impact source descriptions with CSVW.

I propose we go for the simple option for now without including CSVW. @mightymax @wouterbeek do you agree?

**Comment by vemonet** [comment]
So you expect to completely abandon any potential future use of CSVw as `rml:source` in the LDWizard? And move the LDWizard towards a less detailed and less standard solution? (that is not necessarily simpler because it will require to add non standard processes to define the CSV dialect)

With all due respect the simple solution is to include CSVw right now and start building on it :)

CSVw is objectively better than just a string with the file name (because it provides info on the separator for example, but also more). So you will probably want to support CSVw in the future. But right now the proposed solution is that we rewrite part of the LDWizard to make it less good (the RML produced is less detailed). 

If you have the expectation to put back CSVw support in the future... This will add a lot of useless complexity and weird artifacts in the code that will make the code less maintainable in the future.

On the other side the lack of support for CSVw has already been fixed with 5 lines of code, without breaking previous behavior (everyone's win here! Instead of everyone losing features and standards...)

For RocketRML it would require no more than 1 hour to review the quick fix and accept it as it is (I can guide them to do it if needed). Then they can create an issue "improve CSVw support" and improve this support later when it becomes clear. But at least they now knows where they can start working CSVw support to make it better.

**Comment by pmaria** [comment]
> So you expect to completely abandon any potential future use of CSVw as `rml:source` in the LDWizard? 

No, I would introduce it when it is fully supported.

> And move the LDWizard towards a less detailed and less standard solution? 

You could argue that it is more standard, since the RML specification itself doesn't say anything about CSVW. This is something that was added outside of the spec and not described in any detailed specification. Hence, not all mapping engines support CSVW, RocketRML included.

> (that is not necessarily simpler because it will require to add non standard processes to define the CSV dialect)

A partial implementation with no coverage of `csvw:Dialect` would need that as well, right?

> 
> With all due respect the simple solution is to include CSVw right now and start building on it :)

What's simpler? Working without CSVW which already works in the current RocketRML, or waiting for the potential merge of partial CSVW support in RocketRML? :)

> CSVw is objectively better than just a string with the file name (because it provides info on the separator for example, but also more). So you will probably want to support CSVw in the future. But right now the proposed solution is that we rewrite part of the LDWizard to make it less good (the RML produced is less detailed).

I do not disagree that CSVW would be better. But only a partial implementation of CSVW has little value IMO, and waiting for a merge slows down progress.
 
> If you have the expectation to put back CSVw support in the future... This will add a lot of useless complexity and weird artifacts in the code that will make the code less maintainable in the future.

How so? As I see it, all that would need to be done is to change the way the RML mapping is generated. And move some dialect configuration into the mapping. But maybe I'm missing something?
 
> On the other side the lack of support for CSVw has already been fixed with 5 lines of code, without breaking previous behavior (everyone's win here! Instead of everyone losing features and standards...)

I don't see the win if `csvw:Dialect` is not supported. Currently we just have a more deeply nested source location value.
Other aspects, like the delimiter, will still have to be configured in another way until that is the case.

> For RocketRML it would require no more than 1 hour to review the quick fix and accept it as it is (I can guide them to do it if needed). Then they can create an issue "improve CSVw support" and improve this support later when it becomes clear. But at least they now knows where they can start working CSVw support to make it better.

With full CSVW support on RocketRML side there is no discussion. I'm just afraid that it might take a while, since I see no movement on those issues.

In any case. I've stated and hopefully clarified my preference.

I can live with your proposal as long as it gets merged soon and we can expect progress on full CSVW support in the near future. Maybe we need to organize a meeting for this on the RocketRML side?

And curious about the opinion @mightymax and other maintainers.

**Comment by wouterbeek** [comment]
Hi @vemonet  @pmaria @mightymax ,

The LD Wizard WG decided to drop support for CSVW to focus on what the community can currently realize.  There was no other reason for dropping CSVW support than the worry that it would require more development hours than are currently available from the community.

> the effort required to use CSVw support in RocketRML is much lower than the one required for removing CSVw from LDWizard

This assessment places the decision of the WG in a new light.  Maybe the best way to move forward is to discuss this topic during the next WG meeting, present the new assessment there, and see whether the WG comes to a different decision based on the new information that we have before us.

Let me know what you think.

**Comment by EnnoMeijers** [comment]
Hi all, 
As a distant reader I am getting a little puzzled by Wouter's post. So please enlighten me: the current discussion is about (temporarily?) dropping the CSVW support in the RML datasource specification in order to use the default RocketRML library? And not about the earlier discussion and decision to remove the support for CSVW/COW transformations completely ? Or did I get it all wrong?

**Comment by wouterbeek** [comment]
I must have missed the whole point of this discussion then...  I hope that others understand the matter at hand better and make a wise decision.  It would be nice to have a high-level summary that we can share during the next WG.

**Comment by pmaria** [comment]
> Note that if you check the [RML specifications](https://rml.io/specs/rml/#source), an URI is expected to be used for `rml:source`, when our 2 alternatives are a string or a blank node :) 
> > The source (`rml:source`) locates the input data source. It is a [[URI](https://rml.io/specs/rml/#bib-uri)] that represents the data source where the data source is.
> 
> Not sure if there are plans to clarify this

Agreed it is quite unclear. There is work being done to create new specs.

> In any case it is not ideal to use a fork as dependency, so I guess the best would be to discuss with the RocketRML team, and see what would be needed to integrate the CSVw support. And depending on their response see if you need to remove CSVw or not

Agreed. @vemonet could you send a reminder msg on those issues to call attention?

**Comment by EnnoMeijers** [comment]
@pmaria Thanks for clearing this up!

**Comment by vemonet** [comment]
I created a pull request for the CSVw support in RocketRML (https://github.com/semantifyit/RocketRML/pull/41)

Note that as it stands there is another outstanding issue that prevents to use the published version of RocketRML: https://github.com/semantifyit/RocketRML/issues/39 until this error is figured out we will need to use a fork

It seems like the issue surface because webpack tries to resolves an optional dependency (a C++ XML parser) when building the LDWizard. I am not an expert in webpack compilation, but maybe there is an easy fix to better handle the optional dependencies within webpack. I don't know if we could do it just through the LDWizard webpack configuration, or if we need to make change in RocketRML though

**Comment by vemonet** [comment]
Hi everyone (@pmaria @EnnoMeijers @wouterbeek )

We have some good news to start this new year:
1. The issue related to the RocketRML optional import for faster XML parsing failing was fixed by ignoring the failing dependency, thanks to @mightymax and Phillipe 
2. The pull request to support CSVw has been merged to the official RocketRML master branch, which means we will be able to use the next version of RocketRML published to `npm` out of the box 🎉

Currently my fork now uses the official master branch of RocketRML, once they have published the new version of RocketRML to `npm`, I will be able to change the dependency to use the `rocketrml` package directly. And once this is done I will create a pull request to integrate RocketRML processing in the LDWizard master branch

**Comment by vemonet** [comment]
While I was writing the previous message the RocketRML team already published a new version to `npm`! So I updated the dependencies and created the pull request: https://github.com/pldn/LDWizard/pull/89

Once this is merged we can close this issue, thanks everyone!

**Comment by mightymax** [comment]
With the merge of @vemonet 's work this issue can be closed.
